ENH: remove unnecessary root-find from skewnorm

maresb opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#writing-the-commit-message

Depending on your changes, you can skip CI operations and save time and energy: 
http://scipy.github.io/devdocs/dev/contributor/continuous_integration.html#skipping

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
None

#### What does this implement/fix?
Previously the `skew_d` function was inverted using `root_scalar`. However, there is a simple closed-form inverse, so use that instead.

#### Additional information
Reference for inverse function:
<https://en.wikipedia.org/wiki/Skew_normal_distribution#Estimation>

To demonstrate that the inverse function is correct:
```python
np.array_equal(d_skew([-s_max,s_max]), [-1, 1])
np.array_equal(d_skew(np.array([-s_max,s_max])), [-1, 1])

d_vals = np.linspace(-1, 1, 100)
s_vals = np.linspace(-s_max, s_max, 100)
assert_almost_equal(d_vals, d_skew(skew_d(d_vals)), decimal=15)
assert_almost_equal(s_vals, skew_d(d_skew(s_vals)), decimal=15)
```
I'm not sure how I could currently turn these into unit tests as the functions are only defined within the `skewnorm.fit` scope. If desired I could refactor them to be top-level in `scipy.stats._continuous_distns` and add these as tests.

The performance improvement is not as dramatic as I expected:
```python
from scipy.stats import skewnorm

# before
%timeit skewnorm.fit([1,2,3,3,4,4,4,5], method="mm")
673 µs ± 18.3 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

# after
%timeit skewnorm.fit([1,2,3,3,4,4,4,5], method="mm")
616 µs ± 10.1 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
```
This is because the most expensive call is `s = stats.skew(data)`, while `root_scalar` is actually quite fast.

mdhaber commented:
Please confirm that this test would fail if the formula for the increase were wrong.
https://github.com/scipy/scipy/blob/cfe80116aaa145061246b8aec0e98248fbefb835/scipy/stats/tests/test_distributions.py#L3625-L3633

If so, you don't need a new test. If not, write a test along those lines.

maresb commented:
Ah, good call. I added a small error of `1e-7` to the result of `d_skew` and the last line with `assert_allclose` does indeed fail.

As suggested I just added a few checks to make sure fits with MoM and MLE behave as expected under sign change.

mdhaber commented on scipy/stats/tests/test_distributions.py:
```diff
@@ -3755,22 +3755,29 @@ def test_fit(self):
+        # test that MLE and MoM behave as expected under sign changes
+        a6p, loc6p, scale6p = stats.skewnorm.fit(rvs, method='mle')
+        a6m, loc6m, scale6m = stats.skewnorm.fit(-rvs, method='mle')
+        assert_allclose([-a6p, -loc6p, scale6p], [a6m, loc6m, scale6m])
```
Arguably, these should be reversed because I think the latter would be considered the "desired" value.
```suggestion
        assert_allclose([a6m, loc6m, scale6m], [-a6p, -loc6p, scale6p])
```
![image](https://github.com/scipy/scipy/assets/6570539/44f4e472-fd84-4bb9-8beb-55f016297d71)

and the condition is not perfectly symmetric, so we try to be careful about that.

mdhaber commented on scipy/stats/tests/test_distributions.py:
```diff
@@ -3755,22 +3755,29 @@ def test_fit(self):
+        assert_allclose([-a6p, -loc6p, scale6p], [a6m, loc6m, scale6m])
+        a7p, loc7p, scale7p = stats.skewnorm.fit(rvs, method='mm')
+        a7m, loc7m, scale7m = stats.skewnorm.fit(-rvs, method='mm')
+        assert_allclose([-a7p, -loc7p, scale7p], [a7m, loc7m, scale7m])
```
```suggestion
        assert_allclose([a7m, loc7m, scale7m], [-a7p, -loc7p, scale7p])
```

maresb replied:
Thanks a lot for the merge @mdhaber! Very interesting, I wasn't aware of the asymmetry in `assert_allclose`.

mdhaber reviewed (approved):
Minor point to be made, but I'll go ahead and merge.
